File: src/ImageProcessor.cs
using System;
using System.Drawing;
using System.Drawing.Imaging;

namespace PIP
{
  class ImageProcessor
  {
    public const int RANGE_OF_8BITS = 256;

    private Bitmap bitmap;
    private Bitmap grayScaleBitmap;
    private int[] histogram;
    private int[] accumulatedHistogram;
    private int[,] rgbHistogram;

    private int weightedHistogram;

    private int maxHistogram;
    private int maxRgbHistogram;

    private int thresholdValue;
    private int otsuThresholdValue;
    private int entropyThresholdValue;

    private Bitmap binaryBitmap;

    public enum ImageType{
      ORIGIN_IMAGE = 0,
      GRAY_IMAGE,
      BINARY_IMAGE
    }


    /// <summary>
    /// Constructor
    /// </summary>
    /// <param name="bitmap">Bitmap of the image to be processed</param>
    public ImageProcessor(Bitmap bitmap)
    {
      setBitmap(bitmap);
    }

    /// <summary>
    /// Get pixel of given position
    /// </summary>
    /// <param name="x">X position of the pixel</param>
    /// <param name="y">Y position of the pixel</param>
    /// <returns>Color of the pixel</returns>
    public Color getPixel(int x, int y)
    {
      return bitmap.GetPixel(x, y);
    }

    /// <summary>
    /// Get width of the bitmap
    /// </summary>
    /// <returns>Width of the bitmap</returns>
    public int getWidth()
    {
      return bitmap.Width;
    }

    /// <summary>
    /// Get height of the bitmap
    /// </summary>
    /// <returns>Height of the bitmap</returns>
    public int getHeight()
    {
      return bitmap.Height;
    }

    /// <summary>
    /// Get bitmap of original image
    /// </summary>
    /// <returns>Bitmap</returns>
    public Bitmap getBitmap()
    {
      return bitmap;
    }

    /// <summary>
    /// Set bitmap of original image,
    /// histogram and other value is set to be uninitialized
    /// </summary>
    /// <param name="bitmap"></param>
    public void setBitmap(Bitmap bitmap)
    {
      if (bitmap != this.bitmap)
      {
        this.bitmap = bitmap;
        // -1 is the default flag to notify the uninitialized variables
        maxHistogram = -1;
        maxRgbHistogram = -1;
        otsuThresholdValue = -1;
        entropyThresholdValue = -1;
        weightedHistogram = -1;
      }
    }

    /// <summary>
    /// Get gray scale bitmap
    /// </summary>
    /// <returns>Gray scale bitmap</returns>
    public Bitmap getGrayScaleBitmap()
    {
      if (grayScaleBitmap != null)
      {
        return grayScaleBitmap;
      }
      grayScaleBitmap = new Bitmap(bitmap.Width, bitmap.Height);
      /*
      for (int i = 0; i < bitmap.Width; ++i)
      {
        for (int j = 0; j < bitmap.Height; ++j)
        {
          Color color = bitmap.GetPixel(i, j);
          int grayScale = (int)(color.R * 0.3 + color.G * 0.59
            + color.B * 0.11);
          Color grayColor = Color.FromArgb(grayScale, grayScale, grayScale);
          grayScaleBitmap.SetPixel(i, j, grayColor);
        }
      }
      */
      Graphics g = Graphics.FromImage(grayScaleBitmap);
      ColorMatrix colorMatrix = new ColorMatrix(
        new float[][]
        {
          new float[] {0.3f, 0.3f, 0.3f, 0, 0},
          new float[] {0.59f, 0.59f, 0.59f, 0, 0},
          new float[] {0.11f, 0.11f, 0.11f, 0, 0},
          new float[] {0, 0, 0, 1, 0},
          new float[] {0, 0, 0, 0, 1}
        });
      ImageAttributes attributes = new ImageAttributes();
      attributes.SetColorMatrix(colorMatrix);
      g.DrawImage(bitmap, new Rectangle(0, 0, bitmap.Width, bitmap.Height),
        0, 0, bitmap.Width, bitmap.Height, GraphicsUnit.Pixel, attributes);
      g.Dispose();
    
      return grayScaleBitmap;
    }

    /// <summary>
    /// Get histogram array
    /// </summary>
    /// <returns>Histogram array</returns>
    public int[] getHistogram()
    {
      if (histogram != null)
      {
        return histogram;
      }
      histogram = new int[RANGE_OF_8BITS];
      accumulatedHistogram = new int[RANGE_OF_8BITS];

      Bitmap grayScaleBitmap = getGrayScaleBitmap();
      for (int i = 0; i < grayScaleBitmap.Width; ++i)
      {
        for (int j = 0; j < grayScaleBitmap.Height; ++j)
        {
          histogram[grayScaleBitmap.GetPixel(i, j).R]++;
        }
      }

      maxHistogram = histogram[0];
      accumulatedHistogram[0] = histogram[0];
      for (int i = 1; i < RANGE_OF_8BITS; ++i)
      {
        if (histogram[i] > maxHistogram)
        {
          maxHistogram = histogram[i];
        }
        accumulatedHistogram[i] = accumulatedHistogram[i - 1] + histogram[i];
      }
      return histogram;
    }

    /// <summary>
    /// Get accumulated histogram value
    /// </summary>
    /// <returns>Accumulated histogram value</returns>
    public int[] getAccumulatedHistogram()
    {
      if (accumulatedHistogram == null)
      {
        getHistogram();
      }
      return accumulatedHistogram;
    }

    /// <summary>
    /// Get sum(i * histogram[i]) of histogram array
    /// </summary>
    /// <returns>Weighted sum of histogram array</returns>
    int getWeightedHistogram()
    {
      if (weightedHistogram != -1)
      {
        return weightedHistogram;
      }
      // Make sure histogram is calculate
      getHistogram();

      weightedHistogram = 0;
      for (int i = 0; i < RANGE_OF_8BITS; ++i)
      {
        weightedHistogram += i * histogram[i];
      }
      return weightedHistogram;
    }

    /// <summary>
    /// Get RGB hitogram array
    /// </summary>
    /// <returns>RGB hitogram array</returns>
    public int[,] getRgbHistogram()
    {
      if (rgbHistogram != null)
      {
        return rgbHistogram;
      }
      rgbHistogram = new int[RANGE_OF_8BITS, 3];
      for (int i = 0; i < bitmap.Width; ++i)
      {
        for (int j = 0; j < bitmap.Height; ++j)
        {
          rgbHistogram[bitmap.GetPixel(i, j).R, 0]++;
          rgbHistogram[bitmap.GetPixel(i, j).G, 1]++;
          rgbHistogram[bitmap.GetPixel(i, j).B, 2]++;
        }
      }

      maxRgbHistogram = -1;
      int imageSize = bitmap.Height * bitmap.Width;
      for (int i = 0; i < RANGE_OF_8BITS; ++i)
      {
        for (int j = 0; j < 3; ++j)
        {
          if (rgbHistogram[i, j] > maxRgbHistogram)
          {
            maxRgbHistogram = rgbHistogram[i, j];
          }
        }
      }
      return rgbHistogram;
    }

    /// <summary>
    /// Get max histogram value
    /// </summary>
    /// <returns>Max histogram value</returns>
    public int getMaxHistogram()
    {
      if (maxHistogram < 0)
      {
        getHistogram();
      }
      return maxHistogram;
    }

    /// <summary>
    /// Get max RGB histogram value
    /// </summary>
    /// <returns>Max RGB histogram value</returns>
    public int getMaxRgbHistogram()
    {
      if (maxRgbHistogram < 0)
      {
        getRgbHistogram();
      }
      return maxRgbHistogram;
    }

    /// <summary>
    /// Get threshold value calculated by Otsu Algorithm
    /// </summary>
    /// <returns>Threshold value</returns>
    public int getOtsuThresholdValue()
    {
      if (otsuThresholdValue != -1)
      {
        return otsuThresholdValue;
      }
      // Make sure histogram is calculate
      getHistogram();

      // Calculate when threshold value = 0,
      // then calculate on the base of former outcome

      // Sum of pixels with index below thresholdValue 
      int belowSum = 0;
      // Sum of pixels with index upper than threshold value 
      int upperSum = accumulatedHistogram[RANGE_OF_8BITS - 1];

      // Sum of pixels multiplies index value below thresholdValue 
      int belowWeighted = 0;
      // Sum of pixels multiplies index value below threshold value 
      int upperWeighted = getWeightedHistogram();

      // Average of pixels below thresholdValue
      double belowAverage = 0;
      // Average of pixels upper than thresholdValue
      double upperAverage;
      if (upperSum == 0)
      {
        upperAverage = 0;
      }
      else
      {
        upperAverage = (double)upperWeighted / upperSum;
      }

      // Max variance between two parts
      double maxVarianceBetween = belowSum * upperSum
        * (belowAverage - upperAverage) * (belowAverage - upperAverage);
      otsuThresholdValue = 0;

      // Loop threshold value to get max variance between two parts
      for (int thresholdValue = 0; thresholdValue < RANGE_OF_8BITS; ++thresholdValue)
      {
        belowSum += histogram[thresholdValue];
        upperSum -= histogram[thresholdValue];

        if (belowSum == 0)
        {
          continue;
        }
        if (upperSum == 0)
        {
          break;
        }

        belowWeighted += histogram[thresholdValue] * thresholdValue;
        upperWeighted -= histogram[thresholdValue] * thresholdValue;

        belowAverage = (double)belowWeighted / belowSum;
        upperAverage = (double)upperWeighted / upperSum;

        // Variance between below and upper part
        double variaceBetween = (double)belowSum * upperSum
          * (belowAverage - upperAverage) * (belowAverage - upperAverage);

        // Update threshold value if variance between two parts are
        // larger than maxVarianceBetween
        if (variaceBetween > maxVarianceBetween)
        {
          maxVarianceBetween = variaceBetween;
          otsuThresholdValue = thresholdValue;
        }
      }
      return otsuThresholdValue;
    }

    /// <summary>
    /// Get threshold value calculated by Entropy Algorithm
    /// </summary>
    /// <returns>Threshold value</returns>
    public int getEntropyThresholdValue()
    {
      if (entropyThresholdValue != -1)
      {
        return entropyThresholdValue;
      }
      // Make sure histogram is calculate
      getHistogram();

      // Sum of x * log(x) with index below thresholdValue
      double belowSumXLogX = 0;
      // Sum of x * log(x) with index upper than threshold value
      double upperSumXLogX = 0;
      for (int i = 0; i < RANGE_OF_8BITS; ++i)
      {
        if (histogram[i] != 0)
        {
          // Log on base e
          upperSumXLogX += histogram[i] * Math.Log((double)histogram[i]);
        }
      }

      double maxEntropy = 0;

      for (int thresholdValue = 0; thresholdValue < RANGE_OF_8BITS; ++thresholdValue)
      {
        if (histogram[thresholdValue] == 0)
        {
          continue;
        }

        int upperSum = accumulatedHistogram[RANGE_OF_8BITS - 1]
                       - accumulatedHistogram[thresholdValue];
        if (upperSum == 0)
        {
          break;
        }

        double xLogX = histogram[thresholdValue] * Math.Log(histogram[thresholdValue]);
        belowSumXLogX += xLogX;
        upperSumXLogX -= xLogX;

        double belowEntropy = Math.Log(accumulatedHistogram[thresholdValue])
                              - belowSumXLogX / accumulatedHistogram[thresholdValue];
        double upperEntropy = Math.Log(upperSum)
                              - upperSumXLogX / upperSum;
        double entropy = belowEntropy + upperEntropy;

        if (entropy > maxEntropy)
        {
          entropyThresholdValue = thresholdValue;
          maxEntropy = entropy;
        }
      }
      return entropyThresholdValue;
    }

    /// <summary>
    /// Get binary bitmap with given threshold value,
    /// pixels with equal or less value than threshold value will be set 
    /// to be 0 and those with more value to be 255
    /// </summary>
    /// <param name="thresholdValue">Threshold value, -1 for former value</param>
    /// <returns></returns>
    public Bitmap getBinaryBitmap(int thresholdValue = -1)
    {
      if (binaryBitmap != null && thresholdValue == -1)
      {
        return binaryBitmap;
      }

      // If no former threshold value exist, set it to be 0
      if (thresholdValue == -1)
      {
        thresholdValue = 0;
      }
      this.thresholdValue = thresholdValue;

      binaryBitmap = new Bitmap(bitmap);
      Bitmap grayScaleBitmap = getGrayScaleBitmap();

      // Set pixels with value less than threshold value to be 0,
      // and more with 255
      for (int i = 0; i < grayScaleBitmap.Width; ++i)
      {
        for (int j = 0; j < grayScaleBitmap.Height; ++j)
        {
          Color color = grayScaleBitmap.GetPixel(i, j);
          if (color.R <= thresholdValue)
          {
            binaryBitmap.SetPixel(i, j, Color.Black);
          }
          else
          {
            binaryBitmap.SetPixel(i, j, Color.White);
          }
        }
      }
      return binaryBitmap;
    }
  }
}

File: src/ImageProcessor.cpp
#include <QtCore/qmath.h>
#include <QImage>
#include <QPixmap>

#include "ImageProcessor.h"
#include "ImagePolicy.h"

ImageProcessor::ImageProcessor(QString fileName) :
    originImage(0),
    grayScaleImage(0),
    binaryImage(0),
    equalImage(0),
    contrastImage(0),
    brightImage(0),
    isHisCaled(false),
    isRgbHisCaled(false),
    weightedHisSum(0),
    otsuThreshold(0),
    entropyThreshold(0),
    lowerThreshold(0),
    higherThreshold(0),
    isOtsuCaled(false),
    isEntropyCaled(false),
    grayScalePolicy(ImagePolicy::MATCH_LUMINANCE),
    thresholdPolicy(ImagePolicy::OTSU)
{
    setImage(fileName);
}

ImageProcessor::~ImageProcessor()
{
    if (originImage) {
        delete originImage;
    }
    if (grayScaleImage) {
        delete grayScaleImage;
    }
    if (binaryImage) {
        delete binaryImage;
    }
    if (equalImage) {
        delete equalImage;
    }
    if (contrastImage) {
        delete contrastImage;
    }
    if (brightImage) {
        delete brightImage;
    }
}

void ImageProcessor::setImage(QString fileName)
{
    this->fileName = fileName;
    originImage = new QImage(fileName);
    doFormatProcess(originImage);
    resetUncalculated();
}

void ImageProcessor::doFormatProcess(QImage *image)
{
    if (image == 0) {
        return;
    }
    if (image->format() != QImage::Format_RGB32) {
        if (image->format() == QImage::Format_ARGB32) {
            // set pixel with alpha 0 to be white
            uchar* bits = image->bits();
            int size = image->width() * image->height();
            for (int i = 0; i < size; ++i) {
                int alpha = *(bits + 3);
                if (alpha == 0) {
                    *bits = MAX_OF_8BITS;
                    *(bits + 1) = MAX_OF_8BITS;
                    *(bits + 2) = MAX_OF_8BITS;
                } else if (alpha != MAX_OF_8BITS) {
                    // if alpha is not 255, set rgb to be semitransparent
                    *bits = *bits * alpha / MAX_OF_8BITS +
                            MAX_OF_8BITS - alpha;
                    *(bits + 1) = *(bits + 1) * alpha / MAX_OF_8BITS +
                            MAX_OF_8BITS - alpha;
                    *(bits + 2) = *(bits + 2) * alpha / MAX_OF_8BITS +
                            MAX_OF_8BITS - alpha;
                }
                bits += 4;
            }
        }
        *image = image->convertToFormat(QImage::Format_RGB32);
    }
}

QImage* ImageProcessor::getContrastImage(int contrast)
{
    // make sure histogram is calculated
    getHistogram();

    if (contrastImage) {
        delete contrastImage;
    }

    const uchar* originPtr = originImage->constBits();
    int height = originImage->height();
    int width = originImage->width();

    contrastImage = new QImage(width, height, originImage->format());
    uchar* newPtr = contrastImage->bits();

    int averageBrightness;
    int originColor = 0;
    int adjustedColor = 0;

    // compute averageBrightness
    long brightnessSum = 0;
    for (int i = 0; i < RANGE_OF_8BITS; ++i) {
        brightnessSum += histogram[i];
    }
    averageBrightness = brightnessSum / width / height;

    // adjust each pixel
    for (int i = 0; i < width; ++i) {
        for (int j = 0; j < height; ++j) {
            for (int rgb = 0; rgb < 3; ++rgb) {
                originColor = *(originPtr +
                                PIXEL_SIZE * (width * j + i) + rgb);
                adjustedColor = (int)((double)averageBrightness + (double)
                                      (originColor - averageBrightness)
                                      * (double)contrast / 100);
                if (adjustedColor < 0) {
                    adjustedColor = 0;
                }
                if (adjustedColor > MAX_OF_8BITS) {
                    adjustedColor = MAX_OF_8BITS;
                }
                *(newPtr + PIXEL_SIZE * (width * j + i) + rgb) =
                        adjustedColor;
            }
        }
    }
    return contrastImage;
}

QImage* ImageProcessor::getBrightnessImage(int brightness)
{
    if (brightImage) {
        delete brightImage;
    }

    brightImage = new QImage(originImage->size(), QImage::Format_RGB32);
    unsigned char* originPtr = originImage->bits();
    unsigned char* operatedPtr = brightImage->bits();
    int height = originImage->height();
    int width = originImage->width();
    int adjustedColor;

    if (brightness > MAX_OF_8BITS) {
        brightness = MAX_OF_8BITS;
    } else if (brightness < -MAX_OF_8BITS) {
        brightness = -MAX_OF_8BITS;
    }
    for (int i = 0; i < width; ++i) {
        for (int j = 0; j < height; ++j) {
            for (int k = 0; k < 3; ++k) {
                adjustedColor = *(originPtr + 4 * (j * width + i) + k)
                        + brightness;
                if (adjustedColor < 0) {
                    adjustedColor = 0;
                } else if (adjustedColor > MAX_OF_8BITS) {
                    adjustedColor = MAX_OF_8BITS;
                }
                *(operatedPtr + 4 * (j * width + i) + k) = adjustedColor;
            }
        }
    }
    return brightImage;
}

void ImageProcessor::resetUncalculated()
{
    // reset to uncalculated state
    if (grayScaleImage) {
        delete grayScaleImage;
        grayScaleImage = 0;
    }
    if (binaryImage) {
        delete binaryImage;
        binaryImage = 0;
    }
    if (equalImage) {
        delete equalImage;
        equalImage = 0;
    }
    isHisCaled = false;
    isRgbHisCaled = false;
    isOtsuCaled = false;
    isEntropyCaled = false;
}

void ImageProcessor::setGrayScalePolicy(ImagePolicy::GrayScalePolicy policy)
{
    if (grayScalePolicy != policy) {
        grayScalePolicy = policy;
        if (grayScaleImage) {
            delete grayScaleImage;
            grayScaleImage = 0;
        }
        if (binaryImage) {
            delete binaryImage;
            binaryImage = 0;
        }
        isHisCaled = false;
        isOtsuCaled = false;
        isEntropyCaled = false;
    }
}

ImagePolicy::GrayScalePolicy ImageProcessor::getGrayScalePolicy()
{
    return grayScalePolicy;
}

void ImageProcessor::setThresholdPolicy(
        ImagePolicy::ThresholdPolicy policy, int lower, int higher)
{
    thresholdPolicy = policy;
    if (policy == ImagePolicy::COSTUMED) {
        lowerThreshold = lower;
        higherThreshold = higher;
    }
}

ImagePolicy::ThresholdPolicy ImageProcessor::getThresholdPolicy()
{
    return thresholdPolicy;
}

int ImageProcessor::getLowerThreshold()
{
    switch (thresholdPolicy) {
    case ImagePolicy::COSTUMED:
        return lowerThreshold;
    default:
        return 0;
    }
}

int ImageProcessor::getHigherThreshold()
{
    switch (thresholdPolicy) {
    case ImagePolicy::OTSU:
        return getOtsuThreshold();
    case ImagePolicy::ENTROPY:
        return getEntropyThreshold();
    case ImagePolicy::COSTUMED:
        return higherThreshold;
    }
    return 0;
}

int ImageProcessor::getOtsuThreshold()
{
    // lazy calculation
    if (!isOtsuCaled) {
        // Make sure histogram is calculate
        getHistogram();

        // Calculate when threshold value = 0,
        // then calculate on the base of former outcome

        // Sum of pixels with index below thresholdValue
        int belowSum = 0;
        // Sum of pixels with index upper than threshold value
        int upperSum = accHistogram[RANGE_OF_8BITS - 1];

        // Sum of pixels multiplies index value below thresholdValue
        int belowWeighted = 0;
        // Sum of pixels multiplies index value below threshold value
        int upperWeighted = weightedHisSum;

        // Average of pixels below thresholdValue
        double belowAverage = 0;
        // Average of pixels upper than thresholdValue
        double upperAverage;
        if (upperSum == 0)
        {
            upperAverage = 0;
        }
        else
        {
            upperAverage = (double)upperWeighted / upperSum;
        }

        // Max variance between two parts
        double maxVarianceBetween = belowSum * upperSum
                * (belowAverage - upperAverage) * (belowAverage - upperAverage);
        otsuThreshold = 0;

        // Loop threshold value to get max variance between two parts
        for (int thresholdValue = 0; thresholdValue < RANGE_OF_8BITS; ++thresholdValue)
        {
            belowSum += histogram[thresholdValue];
            upperSum -= histogram[thresholdValue];

            if (belowSum == 0)
            {
                continue;
            }
            if (upperSum == 0)
            {
                break;
            }

            belowWeighted += histogram[thresholdValue] * thresholdValue;
            upperWeighted -= histogram[thresholdValue] * thresholdValue;

            belowAverage = (double)belowWeighted / belowSum;
            upperAverage = (double)upperWeighted / upperSum;

            // Variance between below and upper part
            double variaceBetween = (double)belowSum * upperSum
                    * (belowAverage - upperAverage) * (belowAverage - upperAverage);

            // Update threshold value if variance between two parts are
            // larger than maxVarianceBetween
            if (variaceBetween > maxVarianceBetween)
            {
                maxVarianceBetween = variaceBetween;
                otsuThreshold = thresholdValue;
            }
        }
        if (otsuThreshold == 0) {
            otsuThreshold = 1;
        }
        isOtsuCaled = true;
    }
    return otsuThreshold;
}

int ImageProcessor::getEntropyThreshold()
{
    // lazy calculation
    if (!isEntropyCaled) {
        // Make sure histogram is calculate
        getHistogram();

        // Sum of x * log(x) with index below thresholdValue
        double belowSumXLogX = 0;
        // Sum of x * log(x) with index upper than threshold value
        double upperSumXLogX = 0;
        for (int i = 0; i < RANGE_OF_8BITS; ++i)
        {
            if (histogram[i] != 0)
            {
                // Log on base e
                upperSumXLogX += histogram[i] * qLn((double)histogram[i]);
            }
        }

        double maxEntropy = 0;

        for (int thresholdValue = 0; thresholdValue < RANGE_OF_8BITS; ++thresholdValue)
        {
            if (histogram[thresholdValue] == 0)
            {
                continue;
            }

            int upperSum = accHistogram[RANGE_OF_8BITS - 1]
                    - accHistogram[thresholdValue];
            if (upperSum == 0)
            {
                break;
            }

            double xLogX = histogram[thresholdValue] * qLn(histogram[thresholdValue]);
            belowSumXLogX += xLogX;
            upperSumXLogX -= xLogX;

            double belowEntropy = qLn(accHistogram[thresholdValue])
                    - belowSumXLogX / accHistogram[thresholdValue];
            double upperEntropy = qLn(upperSum)
                    - upperSumXLogX / upperSum;
            double entropy = belowEntropy + upperEntropy;

            if (entropy > maxEntropy)
            {
                entropyThreshold = thresholdValue;
                maxEntropy = entropy;
            }
        }
        isEntropyCaled = true;
    }
    return entropyThreshold;
}

QImage* ImageProcessor::getOriginImage()
{
    return originImage;
}

inline uchar ImageProcessor::getGrayValue(
        const uchar* rgb, ImagePolicy::GrayScalePolicy policy)
{
    switch(policy) {
    case ImagePolicy::GREEN_ONLY:
        return *(rgb + 1);
    case ImagePolicy::MATCH_LUMINANCE:
        return ((*rgb) * 28 + *(rgb + 1) * 151 + *(rgb + 2) * 77) >> 8;
    case ImagePolicy::RGB_AVERAGE:
        return (*rgb + *(rgb + 1) + *(rgb + 2)) / 3;
    default:
        return 0;
    }
}

QImage* ImageProcessor::getGrayScaleImage()
{
    // lazy calculation
    if (!grayScaleImage) {
        int width = originImage->size().width();
        int height = originImage->size().height();
        int size = width * height;

        // gray scale image of origin size
        if (grayScaleImage) {
            delete grayScaleImage;
        }
        grayScaleImage = new QImage(width, height, originImage->format());

        // pointer to originBits being processed
        const uchar* originPtr = originImage->constBits();
        // pointer to grayBits being processed
        uchar* grayPtr = grayScaleImage->bits();

        for (int i = 0; i < size; ++i) {
            // calculate gray value according to strategy
            uchar grayValue = getGrayValue(originPtr, grayScalePolicy);
            for (int rgb = 0; rgb < 3; ++rgb) {
                *(grayPtr + rgb) = grayValue;
            }
            // set alpha value to be 255
            *(grayPtr + 3) = MAX_OF_8BITS;
            // move pointer of origin and gray scale pointer to next pixel
            originPtr += PIXEL_SIZE;
            grayPtr += PIXEL_SIZE;
        }
    }
    return grayScaleImage;
}

int* ImageProcessor::getHistogram()
{
    // lazy calculation
    if (!isHisCaled) {
        // make sure gray scale image is calculated
        getGrayScaleImage();
        // init arrays
        for (int i = 0; i < RANGE_OF_8BITS; ++i) {
            histogram[i] = 0;
        }

        // pointer to pixels being processed which will not change
        // pixels in grayImage
        const uchar* grayPtr = grayScaleImage->constBits();

        int size = grayScaleImage->size().width() *
                grayScaleImage->size().height();
        for (int i = 0; i < size; ++i) {
            histogram[*grayPtr]++;
            // point to next pixel
            grayPtr += PIXEL_SIZE;
        }

        // calculate accumulated histogram and weighted sum
        int sum = 0;
        weightedHisSum = 0;
        for (int i = 0; i < RANGE_OF_8BITS; ++i) {
            sum += histogram[i];
            accHistogram[i] = sum;
            weightedHisSum += i * histogram[i];
        }

        isHisCaled = true;
    }
    return histogram;
}

int* ImageProcessor::getRgbHistogram()
{
    // lazy calculation
    if (!isRgbHisCaled) {
        // init arrays
        for (int i = 0; i < RANGE_OF_8BITS; ++i) {
            for (int j = 0; j < 3; ++j) {
                rgbHistogram[i][j] = 0;
            }
        }

        // pointer to pixels being processed which will not change
        // pixels in originImage
        const uchar* originPtr = originImage->constBits();

        int size = originImage->width() * originImage->height();
        for (int i = 0; i < size; ++i) {
            // blue
            rgbHistogram[*originPtr][2]++;
            // green
            rgbHistogram[*(originPtr + 1)][1]++;
            // red
            rgbHistogram[*(originPtr + 2)][0]++;
            // point to next pixel
            originPtr += PIXEL_SIZE;
        }

        isRgbHisCaled = true;
    }
    return *rgbHistogram;
}

QImage* ImageProcessor::getBinaryImage()
{
    // make sure histogram is calculated
    getHistogram();

    int lower = getLowerThreshold();
    int higher = getHigherThreshold();

    int width = grayScaleImage->width();
    int height = grayScaleImage->height();
    int size = width * height;
    if (binaryImage) {
        delete binaryImage;
    }
    binaryImage = new QImage(width, height, originImage->format());

    const uchar* grayPtr = grayScaleImage->constBits();
    uchar* binaryPtr = binaryImage->bits();

    for (int i = 0; i < size; ++i) {
        if (*grayPtr < lower || *grayPtr >= higher) {
            // set two sides of threshold white
            // blue
            *binaryPtr = MAX_OF_8BITS;
            // green
            *(binaryPtr + 1) = MAX_OF_8BITS;
            // red
            *(binaryPtr + 2) = MAX_OF_8BITS;
            // alpha
            *(binaryPtr + 3) = MAX_OF_8BITS;
        } else {
            // blue
            *binaryPtr = 0;
            // green
            *(binaryPtr + 1) = 0;
            // red
            *(binaryPtr + 2) = 0;
            // alpha
            *(binaryPtr + 3) = MAX_OF_8BITS;
        }
        grayPtr += PIXEL_SIZE;
        binaryPtr += PIXEL_SIZE;
    }
    return binaryImage;
}

QImage* ImageProcessor::getEqualImage()
{
    if (equalImage) {
        return equalImage;
    }
    getHistogram();
    getGrayScaleImage();

    // min level of rgb, within [0, 256)
    int minLevel;
    for (int i = 0; i < RANGE_OF_8BITS; ++i) {
        if (histogram[i] > 0) {
            minLevel = i;
            break;
        }
    }
    // max level of rgb, within [0, 256)
    int maxLevel;
    for (int i = RANGE_OF_8BITS - 1; i >= 0; --i) {
        if (histogram[i] > 0) {
            maxLevel = i;
            break;
        }
    }

    equalImage = new QImage(originImage->width(), originImage->height(),
                            originImage->format());
    uchar* bits = equalImage->bits();
    const uchar* oBits = grayScaleImage->constBits();
    int size = equalImage->width() * equalImage->height();
    for (int i = 0; i < size; ++i) {
        uchar value = (double)accHistogram[*oBits] / size * MAX_OF_8BITS;
        for (int rgb = 0; rgb < 3; ++rgb) {
            *bits = value;
            ++bits;
            ++oBits;
        }
        ++bits;
        ++oBits;
    }
    return equalImage;
}
